ASSERTION FAILED: iterator != m_sessionToPageIDsMap.end() in WebProcessPool::pageEndUsingWebsiteDataStore

szewai opened the pull request:
#### 1ba1a365ee238ffec2a6c8239a3315d8d747cc09
<pre>
ASSERTION FAILED: iterator != m_sessionToPageIDsMap.end() in WebProcessPool::pageEndUsingWebsiteDataStore
<a href="https://rdar.apple.com/162715518">rdar://162715518</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=301142">https://bugs.webkit.org/show_bug.cgi?id=301142</a>

Reviewed by Chris Dumez.

WebPageProxy::getInformationFromImageData() should avoid invoking ensureRunningProcess() if page has been closed;
otherwise, WebProcessPool::pageEndUsingWebsiteDataStore() might be invoked twice on the same page (one from
WebPageProxy::close() and one from WebPageProxy::launchProcess()), and the second call might cause assertion failure.

This patch also fixes other call sites of ensureRunningProcess() and adds an assertion in ensureRunningProcess().

API test: WebKit.GetInformationFromImageDataAfterClosingWebView

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getInformationFromImageData):
(WebKit::WebPageProxy::createIconDataFromImageData):
(WebKit::WebPageProxy::decodeImageData):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::ensureRunningProcess):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::setImmersiveSessionState):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
(TestWebKitAPI::TEST(WebKit, GetInformationFromImageDataAfterClosingWebView)):

Canonical link: <a href="https://commits.webkit.org/301928@main">https://commits.webkit.org/301928@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3d260dfa449dc02c7297f012a27b8a3bdb8068a9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127515 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47163 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38301 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134591 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/79071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/03b72130-3546-4bc2-bcd8-d14592bce4e8) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129387 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47780 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55688 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97064 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/79071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9e77b9a9-5056-4c48-a5bd-4504c6379a1f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130463 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/38216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114199 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77545 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/dc3de7b6-9c62-4c6f-a3cd-5f3cb2cb6ada) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/37043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/32300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77965 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/108076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32728 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137076 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54174 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/41744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/105589 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54685 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110557 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105241 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26829 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50756 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/29206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51754 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54111 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60198 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/53345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/56802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55104 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2058,7 +2058,10 @@ void WebPageProxy::addPlatformLoadParameters(WebProcessProxy&, LoadParameters&)
+    if (m_isClosed)
+        WEBPAGEPROXY_RELEASE_LOG_ERROR(Loading, "ensureRunningProcess: Page is closed");
+
+    if (m_legacyMainFrameProcess->wasTerminated() && !m_isClosed)
```
Why is this no longer relying on `hasRunningProcess()`?

Also, technically, this is a bug at the call site. Can we fix the call site instead of add a `!m_isClosed` assertion here? It makes no sense to ask for a process after a WKWebView has been closed.

szewai replied:
`hasRunningProcess()` returns false if `m_isClosed` is true; here we don't want to relaunch process if `m_isClosed` is true.

> Also, technically, this is a bug at the call site. Can we fix the call site instead of add a !m_isClosed assertion here? It makes no sense to ask for a process after a WKWebView has been closed.

Yes I thought about that and figure this might be more robust. If there is another API like `WebPageProxy::getInformationFromImageData` is added and the developer does not know to check page close condition, we might notice the issue until someone hits the assertion in page close case, while returning a terminated process will just lead the message sending to fail and we return error to API.

But if you think it's better add the page close check at all call sites and assert here I could change the implementation.

cdumez replied:
I don't like that it hides bugs in callers. As evidence by this PR where you don't even fix the call sites, people will stop caring and do wrong things because it kind of works.

szewai replied:
Ok I updated the PR as recommended.

szewai committed:
ASSERTION FAILED: iterator != m_sessionToPageIDsMap.end() in WebProcessPool::pageEndUsingWebsiteDataStore
rdar://162715518
https://bugs.webkit.org/show_bug.cgi?id=301142

Reviewed by Chris Dumez.

WebPageProxy::getInformationFromImageData() should avoid invoking ensureRunningProcess() if page has been closed;
otherwise, WebProcessPool::pageEndUsingWebsiteDataStore() might be invoked twice on the same page (one from
WebPageProxy::close() and one from WebPageProxy::launchProcess()), and the second call might cause assertion failure.

This patch also fixes other call sites of ensureRunningProcess() and adds an assertion in ensureRunningProcess().

API test: WebKit.GetInformationFromImageDataAfterClosingWebView

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getInformationFromImageData):
(WebKit::WebPageProxy::createIconDataFromImageData):
(WebKit::WebPageProxy::decodeImageData):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::ensureRunningProcess):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::setImmersiveSessionState):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
(TestWebKitAPI::TEST(WebKit, GetInformationFromImageDataAfterClosingWebView)):

Canonical link: https://commits.webkit.org/301928@main
